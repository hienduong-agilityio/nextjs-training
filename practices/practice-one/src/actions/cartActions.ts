'use server';

// Libraries
import { revalidatePath } from 'next/cache';

// Services
import { addToCart, deleteFromCart } from '@/services';

// Constants
import { ROUTE, TOAST_MESSAGES } from '@/constants';

/**
 * Server action to add a product to the cart.
 * @param userId - The ID of the user
 * @param productId - The ID of the product
 * @param quantity - Quantity to add
 * @param maxQuantity - Max quantity to add
 * @returns {boolean} - Success or failure of the operation
 */
export async function handleAddToCart({
  userId,
  productId,
  quantity,
  maxQuantity,
}: {
  userId: number;
  productId: string;
  quantity: number;
  maxQuantity: number;
}): Promise<boolean> {
  try {
    const { success } = await addToCart(userId, {
      productId,
      quantity,
      maxQuantity,
    });

    if (success) {
      revalidatePath(ROUTE.CART);
      revalidatePath(ROUTE.ROOT);
    }

    return success;
  } catch (error) {
    throw new Error(TOAST_MESSAGES.API_ERROR);
  }
}

/**
 * Server action to remove a product from the cart.
 * @param userId - The ID of the user
 * @param productId - The ID of the product to remove
 * @returns {boolean} - Success or failure of the operation
 */
export async function handleRemoveFromCart({
  userId,
  productId,
}: {
  userId: number;
  productId: string;
}): Promise<boolean> {
  try {
    const { success } = await deleteFromCart(userId, productId);

    if (success) {
      revalidatePath(ROUTE.CART);
      revalidatePath(ROUTE.ROOT);
    }

    return success;
  } catch (error) {
    throw new Error(TOAST_MESSAGES.API_ERROR);
  }
}
