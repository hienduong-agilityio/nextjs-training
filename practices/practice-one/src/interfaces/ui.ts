export interface IInputGroupProps {
  placeholder?: string;
  value?: string;
  buttonText?: string;
  inputName?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  startIcon?: React.ReactNode;
  customClass?: {
    container?: string;
    inputContainer?: string;
    input?: string;
    button?: string;
  };
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
}
