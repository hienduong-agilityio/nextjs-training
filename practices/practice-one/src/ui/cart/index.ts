export * from './CartTable';
export * from './OrderSummary';
export * from './VoucherInput';
export * from './EmptyCart';

export { default as CartItemRow } from './CartItemRow';
