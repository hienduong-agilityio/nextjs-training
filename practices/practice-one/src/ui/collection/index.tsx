export * from './LoadingCategoryGroup';
export * from './LoadingFilterSortBar';
