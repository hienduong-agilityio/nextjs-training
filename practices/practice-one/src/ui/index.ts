export * from './products/index';
export * from './collection/index';
