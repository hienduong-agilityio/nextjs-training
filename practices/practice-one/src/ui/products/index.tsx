export * from './CartAndFavoriteActions';
export * from './ProductImages';
export * from './ProductInfo';
export { default as QuantityControl } from './QuantityControl';
export * from './SocialShare';
export * from './RelatedProducts';
export * from './ProductDetailTabs';
export * from './LoadingRelatedProducts';
export * from './LoadingProductCard';
export * from './LoadingProductTabs';
export * from './ProductReview';
