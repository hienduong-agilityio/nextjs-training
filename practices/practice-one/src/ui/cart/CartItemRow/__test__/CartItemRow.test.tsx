import {
  render,
  screen,
  fireEvent,
  waitFor,
  act,
} from '@testing-library/react';

import { handleUpdateProductInCart } from '@/actions';
import { CartItemRow } from '@/ui';

jest.mock('@/actions', () => ({
  handleDeleteProductFromCart: jest.fn(),
  handleUpdateProductInCart: jest.fn(),
}));

jest.mock('@/stores', () => ({
  ToastStore: jest.fn(() => ({
    showToast: jest.fn(),
  })),
}));

const mockProps = {
  id: '123',
  thumbnail: '/images/image-placeholder.svg',
  quantity: 2,
  title: 'Test Product',
  price: 50,
};

describe('CartItemRow', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    jest.useFakeTimers();
  });

  afterEach(() => {
    jest.useRealTimers();
  });

  it('should match snapshot', () => {
    const { container } = render(
      <table>
        <tbody>
          <CartItemRow total={0} {...mockProps} />
        </tbody>
      </table>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should revert optimistic quantity on API error during quantity update', async () => {
    (handleUpdateProductInCart as jest.Mock).mockRejectedValue(
      new Error('API error'),
    );

    render(
      <table>
        <tbody>
          <CartItemRow total={0} {...mockProps} />
        </tbody>
      </table>,
    );

    const decrementButton = screen.getByRole('button', { name: /decrement/i });
    const quantityInput = screen.getByRole('spinbutton', {
      name: /quantity input/i,
    });

    expect(quantityInput).toHaveValue(2);

    fireEvent.click(decrementButton);

    expect(quantityInput).toHaveValue(1);

    await act(async () => {
      jest.advanceTimersByTime(1000);
    });

    await waitFor(() => {
      expect(handleUpdateProductInCart).toHaveBeenCalledWith({
        userId: expect.any(Number),
        productId: mockProps.id,
        newQuantity: 1,
      });
    });
  });
});
