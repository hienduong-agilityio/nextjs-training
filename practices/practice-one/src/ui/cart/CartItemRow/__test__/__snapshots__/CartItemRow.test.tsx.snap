// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItemRow should match snapshot 1`] = `
<div>
  <table>
    <tbody>
      <tr>
        <td
          class="flex items-center px-4 py-8 w-max"
        >
          <button
            aria-label="Remove product"
            class="p-2 mr-4 text-danger-300 "
            type="button"
          >
            Ã—
          </button>
          <img
            alt="Test Product"
            class="rounded border w-[90px] h-[60px] object-contain bg-secondary-50"
            data-nimg="1"
            decoding="async"
            height="60"
            loading="lazy"
            src="/images/image-placeholder.svg"
            style="color: transparent;"
            width="90"
          />
          <a
            class="ml-4 hover:underline "
            href="/product/123"
          >
            Test Product
          </a>
        </td>
        <td
          class="px-4 py-8 text-start"
        >
          $
          50
        </td>
        <td
          class="flex items-center justify-start w-1/4 px-4 py-8 space-x-2"
        >
          <div
            aria-label="Quantity Control"
            class="rounded-md gap-4 bg-secondary-50"
          >
            <div
              class="flex"
            >
              <button
                aria-label="decrement"
                class="px-5 py-3 text-primary-100 hover:text-gray-800 disabled:cursor-not-allowed disabled:text-gray-800"
                type="button"
              >
                -
              </button>
              <input
                aria-label="Quantity input"
                class="w-12 text-center border-0 bg-gray-50 focus:ring-0"
                type="number"
                value="2"
              />
              <button
                aria-label="increment"
                class="px-5 py-3 text-primary-100 hover:text-gray-800 disabled:cursor-not-allowed disabled:text-gray-800"
                type="button"
              >
                +
              </button>
            </div>
          </div>
        </td>
        <td
          class="px-4 py-8 text-start"
        >
          $
          100.00
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
